pkg/auth/auth.go

```diff
@@ -0,0 +1,264 @@
+	var auth autorest.Authorizer
+	var err error
+
+	var token *adal.ServicePrincipalToken
+	if c.AuthConfig.UseUserIdentity {
+		token, err = adal.NewServicePrincipalTokenFromManagedIdentity(
+			resource, &adal.ManagedIdentityOptions{ClientID: c.AuthConfig.UserIdentityClientId})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		oauthConfig, err := adal.NewOAuthConfig(
+			c.Cloud.ActiveDirectoryAuthorityHost, c.AuthConfig.TenantID)
+		if err != nil {
+			return nil, err
+		}
+		token, err = adal.NewServicePrincipalToken(
+			*oauthConfig, c.AuthConfig.ClientID, c.AuthConfig.ClientSecret, resource)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	auth = autorest.NewBearerAuthorizer(token)
+	return auth, err
+}
+
+// SetAuthConfig sets the configuration needed for Authentication
+func (c *Config) SetAuthConfig() error {
+	var err error
+	c.Cloud = cloud.AzurePublic
+
+	if authFilepath := os.Getenv("AZURE_AUTH_LOCATION"); authFilepath != "" {
+		auth := &Authentication{}
+		err = auth.NewAuthenticationFromFile(authFilepath)
+		if err != nil {
+			return err
+		}
+		c.AuthConfig = auth
+	}
+
+	if acsFilepath := os.Getenv("ACS_CREDENTIAL_LOCATION"); acsFilepath != "" {
+		c.AcsCredential, err = NewAcsCredential(acsFilepath)
+		if err != nil {
+			return err
+		}
+
+		if c.AcsCredential != nil {
+			var clientId string
+			if !strings.EqualFold(c.AcsCredential.ClientID, "msi") {
+				clientId = c.AcsCredential.ClientID
+			}
+
+			//Set Azure cloud environment
+			c.Cloud = getCloudConfiguration(c.AcsCredential.Cloud)
+			c.AuthConfig = NewAuthentication(
+				clientId,
+				c.AcsCredential.ClientSecret,
+				c.AcsCredential.SubscriptionID,
+				c.AcsCredential.TenantID,
+				c.AcsCredential.UserAssignedIdentityID)
+		}
+
+		if clientID := os.Getenv("AZURE_CLIENT_ID"); clientID != "" {
+			c.AuthConfig.ClientID = clientID
+		}
+
+		if clientSecret := os.Getenv("AZURE_CLIENT_SECRET"); clientSecret != "" {
+			c.AuthConfig.ClientSecret = clientSecret
+		}
+
+		if userIdentityClientId := os.Getenv("VIRTUALNODE_USER_IDENTITY_CLIENTID"); userIdentityClientId != "" {
+			c.AuthConfig.UserIdentityClientId = userIdentityClientId
+		}
+		c.AuthConfig.UseUserIdentity = (len(c.AuthConfig.ClientID) == 0)
+
+		if c.AuthConfig.UseUserIdentity {
+			if len(c.AuthConfig.UserIdentityClientId) == 0 {
+				return fmt.Errorf("neither AZURE_CLIENT_ID or VIRTUALNODE_USER_IDENTITY_CLIENTID is being set")
+			}
+
+			log.G(context.TODO()).Info("Using user identity for Authentication")
+		}
+
+		if tenantID := os.Getenv("AZURE_TENANT_ID"); tenantID != "" {
+			c.AuthConfig.TenantID = tenantID
+		}
+
+		if subscriptionID := os.Getenv("AZURE_SUBSCRIPTION_ID"); subscriptionID != "" {
+			c.AuthConfig.SubscriptionID = subscriptionID
+		}
+	}
+
+	resource := c.Cloud.Services[cloud.ResourceManager].Endpoint
+	options := &arm.ClientOptions{}
+
+	if cs, ok := options.Cloud.Services[cloud.ResourceManager]; ok {
```

Comment:
In line 140, I think options are empty, so will line 142 be meaningless? You can defer the cleanup if this change only contains code move.

Reply:
For some reason, if we didn't add line 140, the `options.Cloud.Services[cloud.ResourceManager]` will return an empty string!

Reply:
Removed and replaced by implicit import for arm package.